migration: Support API migration for muted topics

preetmishra opened the pull request:
This reflects the expected response change for muted_topics, from `[stream_name, topic]` to `[stream_name, topic, date_muted]` in Zulip version 3.0, Feature level 1 (see https://github.com/zulip/zulip/commit/9340cd1a0b7af7af9b83a622e37d02f1a2340609).

Tests amended.

preetmishra commented:
Logged #745 to track related improvements.

preetmishra commented:
@neiljp Thanks for the review. I have updated the PR as per our discussion in the stream. This is larger than what we started with, but I believe the refactoring is worthwhile.

I have also added the proposed extension in #745.

sumanthvrao commented on zulipterminal/model.py:
```diff
@@ -423,7 +427,14 @@ def is_in_muted_topics(self, stream_id: int, topic: str) -> bool:
+        # The intent is to use the following data structure locally for lookups
+        # without having server version & feature level checks throughout the
+        # codebase.
+        muted_topics_lookup = [
```
This is a nice idea to unify both versions! 👍

neiljp commented on zulipterminal/ui_tools/buttons.py:
```diff
@@ -242,8 +242,8 @@ def __init__(self, stream_id: int, topic: str,
                          count=count)
 
-        topic_pair = [self.stream_name, self.topic_name]
-        if topic_pair in controller.model.muted_topics:
```
I'll emphasize that we should add a leading underscore to `Model.muted_topics` - it'll mark it as 'private' (by convention at least), and any hidden uses of it are made evident too :)

neiljp commented on tests/ui/test_utils.py:
```diff
@@ -83,8 +82,23 @@ def test_is_muted(mocker, msg, narrow, muted_streams, muted_topics, muted):
     model.is_muted_stream = (
         mocker.Mock(return_value=(msg.get('stream_id', '') in muted_streams))
     )
+    model.stream_dict = {
+        1: {
+            'name': 'foo',
+        },
+        2: {
+            'name': 'boo',
+        },
+        3: {
+            'name': 'zoo',
+        },
+    }
```
Would using an existing stream fixture make sense here?

preetmishra replied:
The stream names used in its parametrize are different from `stream_dict`. I could change the stream names in a separate commit.

neiljp commented on tests/ui/test_utils.py:
```diff
@@ -83,8 +82,23 @@ def test_is_muted(mocker, msg, narrow, muted_streams, muted_topics, muted):
+        },
+    }
+    if msg['type'] == 'stream':
+        model.is_in_muted_topics = mocker.Mock(return_value=(
+            [model.stream_dict[msg['stream_id']]['name'],
+             msg.get('subject', '')] in muted_topics
+        ))
```
This seems strange since it's basically reimplementing the function?

neiljp commented on tests/model/test_model.py:
```diff
@@ -1516,12 +1516,20 @@ def test_is_muted_stream(self, muted_streams, stream_id, is_muted,
-            ['Stream 2', 'muted topic'],
-            ['Stream 1', 'muted stream muted topic'],
-        ]
+        server_version, server_feature_level = zulip_version
+        if server_version == '3.0' and server_feature_level >= 1:
+            model.muted_topics = [
+                ['Stream 2', 'muted topic', 1530129122],
+                ['Stream 1', 'muted stream muted topic', 1530129122],
+            ]
+        elif server_version == '2.1':
+            model.muted_topics = [
+                ['Stream 2', 'muted topic'],
+                ['Stream 1', 'muted stream muted topic'],
+            ]
```
I'd prefer that we had this version-dependent data in a fixture, if possible. Also can we depend upon just the feature level?

neiljp commented on zulipterminal/model.py:
```diff
@@ -423,7 +427,14 @@ def is_in_muted_topics(self, stream_id: int, topic: str) -> bool:
+        # without having server version & feature level checks throughout the
+        # codebase.
+        muted_topics_lookup = [
+            [stream_name, topic]
+            for stream_name, topic, *_ in self.muted_topics
+        ]
+        return topic_to_search in muted_topics_lookup
```
This approach does work, but it would be beneficial to have some defensive programming (assert on type of the incoming data), and I wonder if we can generate and use just one data-structure, like a dict?

neiljp commented on zulipterminal/model.py:
```diff
@@ -115,8 +115,12 @@ def __init__(self, controller: Any) -> None:
         (self.stream_dict, self.muted_streams,
          self.pinned_streams, self.unpinned_streams) = stream_data
 
+        # NOTE: The expected response has been upgraded from
+        # [stream_name, topic] to [stream_name, topic, date_muted] in
+        # 3.0, Feature level 1.
```
It'd be useful to standardize our notes in the code for this; we could add that to the development section - this would make it easier to check for changes at different server versions and feature levels :)

preetmishra commented:
@neiljp The first two commits make modification in the existing code to facilitate the migration. The last commit is the migration commit (should we call it a bugfix?). Note that the first two commits do not introduce any 'new' feature.

preetmishra commented on zulipterminal/model.py:
```diff
@@ -417,9 +417,10 @@ def fetch_all_topics(self, workers: int) -> None:
-    def is_muted_topic(self, stream_id: int, topic: str) -> bool:
-        if stream_id in self.muted_streams:
-            return True
+    def is_in_muted_topics(self, stream_id: int, topic: str) -> bool:
+        """
+        Returns True if the topic_name is muted via muted_topics.
+        """
```
The reason why I pivoted from `is_muted_topic` is that it is slightly confusing given that we could call a topic muted even if it belongs to a muted stream (?) and in this case, we are more interested in whether the topic belongs to the `muted_topics` data structure.

preetmishra commented:
@neiljp Thanks for the review. In the latest update, I have simplified the parameterize blocks to avoid unnecessary mocking wherever feasible, made `muted_topics` protected with an underscore, added `_muted_topics_dict` along with an assert, and addressed other in-line suggestions (excluding the standard comment that we are discussing in [zulip-terminal > Comments for server versions](https://chat.zulip.org/#narrow/stream/206-zulip-terminal/topic/Comments.20for.20server.20versions)).

preetmishra commented:
(Rebased origin/master.)

preetmishra commented:
(Updated to rebase onto upstream/master.)

preetmishra commented:
Updated to resolve conflicts.

neiljp commented on zulipterminal/model.py:
```diff
@@ -116,8 +116,20 @@ def __init__(self, controller: Any) -> None:
+            (stream_name, topic): (None if self.server_feature_level is None
+                                   else date_muted[0])
+            for stream_name, topic, *date_muted in self._muted_topics
+        }
```
I'm glad you understood what I was suggesting here - this looks much simpler :+1:

I'm not sure we still need to store `self._muted_topics` though? Can we just name the final variable that instead?
(I know we have `_dict` on other variables, but including the type in the name is often superfluous)

neiljp commented on tests/conftest.py:
```diff
@@ -752,6 +752,25 @@ def stream_dict(streams_fixture):
+        'server_feature_level:None',
+    ]
+)
+def muted_topics_dict(request):
```
This is the parametrization of muted_topics in the Model only - we should likely have one for initial data from server, and a test for initial data -> model data, which we don't have? (that was in the previous iteration?)

preetmishra replied:
I didn't have that test in the previous iteration. Thanks for pointing that out.

I have added the test in the latest update with a local `parametrize` block in `Model.py`. It wouldn't be needed in `conftest` as we'd only test against `processed_muted_topics` everywhere else.

neiljp replied:
I'm actually thinking we might move the server response fixtures into a separate fixture file specific to the model in a similar way. What do you think?

preetmishra replied:
The ones that are already in `conftest.py` or also the ones that are used in `test_model.py`?

preetmishra replied:
In any case, I believe it is reasonable.

preetmishra commented on tests/model/test_model.py:
```diff
@@ -1585,7 +1585,7 @@ def test_is_muted_stream(self, muted_streams, stream_id, is_muted,
     def test_is_muted_topic(self, topic, is_muted, stream_dict, model):
         model.stream_dict = stream_dict
-        model.muted_topics = [
+        model._muted_topics = [
```
Hmm. I am unable to find any other `muted_topics` instance in the code. Could you point out where?

Re commit, I am inclined towards having it here since we replace the old references with `is_muted_topic()` and it is easier to follow with less number of changes.

neiljp replied:
I can't find another instance now, so this looks good :+1:

preetmishra commented on tests/ui/test_ui_tools.py:
```diff
@@ -2677,22 +2679,25 @@ def test_init_calls_top_button(self, mocker, width, count, title,
+        ('Django', 'topic2', False, False),
+        ('GSoC', 'topic1', False, False),
     ], ids=[
+        # Assuming 'Django', 'topic1' is muted via muted_topics.
```
Yes, it is needed here to understand why the values are set to `True`/`False` for `is_muted_topic_return_value`. Though, I have added an extra refactor commit to simplify it.

The `is_muted` util test is simplified in this commit itself. We might want to discuss `test_classify_unread_counts` - how would we like it to be handled: Maybe as a separate issue.

preetmishra replied:
Yes, it is needed here to understand why the values are set to `True`/`False` for `is_muted_topic_return_value`. Though, I have added an additional refactor commit to simplify it.

The `is_muted` util test is simplified in this commit itself. We might want to discuss `test_classify_unread_counts` - how would we like it to be handled? Maybe as a separate issue.

preetmishra commented:
@neiljp Thanks for the review. This is ready with improved assertion against server response, improved tests, an additional refactor commit and other in-line suggestions that you mentioned.

neiljp commented:
@preetmishra Thanks for this bugfix and compatibility change (as well as the refactors) :+1: As per PMs I've merged the first four :tada:

I think we can save the last commit for a separate set of changes to review separately. Other than unblocking the other PR, this also could be followed by:
* a server-response fixture refactor (after discussion)
* feature-level comments (we briefly discussed already)
